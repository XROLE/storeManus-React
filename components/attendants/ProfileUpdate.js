import React, { Component, Fragment } from 'react';
import jwtDecode from 'jwt-decode';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import Spinner from 'react-spinkit';
import { decode } from 'punycode';
import { updateProfile } from '../../store/actions/actions';
import AttendantNavigator from './AttendantNavigator';
import Navbar from '../navbar/Navbar';

class ProfileUpdate extends Component {
  constructor(props) {
    super();
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      phoneno: '',
      password: '',
      confirmpassword: '',
      profilepics: '',
      gender: '',
      id: '',
      shouldShowError: true,
      shouldRedirect: true,
    };
  }

  componentDidMount() {
    const token = localStorage.getItem('accessToken');
    const decoded = jwtDecode(token);
    const firstName = decoded.firstname;
    const lastName = decoded.lastname;
    const { email, id } = decoded;

    this.setState({
      firstName,
      lastName,
      email,
      id,
    });
  }


  onChange(e) {
    return this.setState({
      [e.target.id]: e.target.value,
    });
  }

  showFileUrl() {
    const fileSize = document.querySelector('input[type=file]').files[0].size;
    const reader = new FileReader();

    if (fileSize > 70000) {
      return alert('file size is too large. files sized must not exceed 7KB ', fileSize);
    }
    reader.readAsDataURL(document.getElementById('profilepics').files[0]);
    reader.onload = () => {
      this.setState({ profilepics: reader.result });
    };
  }

  handleSubmit(e) {
    e.preventDefault();
    const { update } = this.props;
    const { id } = this.state;
    update(updateProfile(this.state, id));
  }

  notify(error) {
    toast.error(error);
  }

  render() {
    const {
      firstName, lastName, email, phoneNumber, password, confirmpassword, shouldShowError, shouldRedirect,
    } = this.state;

    const { pending, updateError, success } = this.props;

    if (success === true && shouldRedirect === true) {
      localStorage.removeItem('accessToken');
      this.setState({ shouldRedirect: false });
      return <Redirect to="/login" />;
    }

    if (updateError && shouldShowError) {
      this.notify(updateError);
      this.setState({
        shouldShowError: false,
      });
    }

    return (
      <Fragment>
        <Navbar />
        <AttendantNavigator />
        <div className="attendant-display-functionality-div">
          <div>
            <form action="" id="att-pro-update-form">
              <div>
                <p>UPDATE PROFILE</p>
              </div>
              <div>
                <input type="text" id="firstName" placeholder="First Name" value={firstName} readOnly />
              </div>
              <div>
                <input type="text" id="lastName" placeholder="Last Name" value={lastName} readOnly />
              </div>
              <div>
                <input type="email" id="email" placeholder="Email" value={email} readOnly />
              </div>
              <div>
                <input type="number" id="phoneno" placeholder="Phone Number" onChange={e => this.onChange(e)} value={phoneNumber} />
              </div>
              <div>
                <input type="password" name="password" id="password" placeholder="Password" onChange={e => this.onChange(e)} value={password} />
              </div>
              <div>
                <input type="password" name="confirmpassword" id="confirmpassword" placeholder="Confirm Password" onChange={e => this.onChange(e)} value={confirmpassword} />
              </div>
              <div className="att-pro-pix-div">
                <div>
                  <label htmlFor="" style={{ color: 'black' }}>Gender: </label>
                  <select name="gender" id="gender" style={{ padding: 'none', color: 'black' }} onChange={e => this.onChange(e)}>
                    <option name="gender">--Select Gender--</option>
                    <option name="gender"> Male </option>
                    <option name="gender"> Female </option>
                  </select>
                </div>

                <input
                  type="file"
                  id="profilepics"
                  onChange={() => {
                    this.showFileUrl();
                  }}
                />

              </div>
              <div>
                <button className="link" id="updateProfileButton" type="submit" onClick={e => this.handleSubmit(e)}>UPDATE</button>
              </div>
              { pending === true && (
              <Spinner name="circle" className="spinner" id="reactLoader" />
              )}
            </form>
            <ToastContainer />
          </div>
        </div>
      </Fragment>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return ({
    update: profile => dispatch(profile),
  });
}
const mapStateToProps = state => ({
  token: state.profile.token,
  pending: state.profile.pending,
  updateError: state.profile.error,
  success: state.profile.success,
});

export default connect(mapStateToProps, mapDispatchToProps)(ProfileUpdate);
