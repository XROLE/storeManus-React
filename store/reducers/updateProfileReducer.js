import {
  UPDATE_IN_PROGRESS,
  UPDATE_ERROR,
  UPDATE_SUCCESS,
} from '../actions/actionTypes';

const initialState = {
  pending: false,
  error: '',
  success: false,
  message: '',
};

export default function (state = initialState, action) {
  switch (action.type) {
    case UPDATE_IN_PROGRESS: {
      return {
        ...state,
        pending: true,
      };
    }
    case UPDATE_SUCCESS: {
      return {
        ...state,
        pending: false,
        success: action.payload.success,
        message: action.payload.message,
      };
    }
    case UPDATE_ERROR: {
      return {
        ...state,
        pending: false,
        error: action.error,
      };
    }
    default:
      return state;
  }
}
