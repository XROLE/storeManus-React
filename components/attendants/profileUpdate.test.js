import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { BrowserRouter } from 'react-router-dom';
import axios from 'axios';
import ProfileUpdate from './ProfileUpdate';
import updateProfileReducer from '../../store/reducers/updateProfileReducer';

jest.mock('axios');
const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MzAsImZpcnN0bmFtZSI6InJhamkiLCJsYXN0bmFtZSI6ImVtbXkiLCJlbWFpbCI6InJhamlAZW1teS5jb20iLCJwYXNzd29yZCI6IiQyYiQxMCRvL0MxQlNsd1hKd3IwRlp1ekdYN2V1ZmlsTFZJZzBsYmpmS3AzWHhJQ1c3R2subkJpOFdmTyIsInBob25lbm8iOiI4OCIsImdlbmRlciI6IkZlbWFsZSIsInByb2ZpbGVwaWNzIjoiaHR0cHM6Ly9yZXMuY2xvdWRpbmFyeS5jb20veHJvbGUvaW1hZ2UvdXBsb2FkL3YxNTUyMDUwOTc2L3QydGFyaWpjd3owbXVrY3ZkcWFtLmpwZyIsImlhdCI6MTU1MjI4OTcyNn0.JDf4zERSlM3vepD_9bxCIwMS4JBGYA859bEHqjMycTs';

class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

global.localStorage = new LocalStorageMock();
localStorage.setItem('accessToken', token);

describe('Profile component', () => {
  it('Should render just the Attendants component', () => {
    const component = shallow(
      <ProfileUpdate />,
    );
    expect(component.exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});

describe('Update Profile Component', () => {
  const update = jest.fn();
  const initialState = {
    profile: {
      pending: false,
      error: '',
      success: false,
      message: '',
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(() => initialState);
  const dispatchMock = jest.spyOn(store, 'dispatch');
  const pending = false;
  const updateError = '';
  const success = false;

  const props = {
    profile: {
      update,
      pending,
      updateError,
      token,
      success,
    },
  };

  const wrapper = mount(
    <Provider store={store}>
      <BrowserRouter>
        <ProfileUpdate {...props} />
      </BrowserRouter>
    </Provider>,
  );

  it('should call the update Profile function', () => {
    wrapper.find('#phoneno').simulate(
      'change', {
        target: {
          id: 'phoneno', value: '07087653446',
        },
      },
    );

    wrapper.find('#password').simulate(
      'change', {
        target: {
          id: 'password', value: 'sameWhiteLady',
        },
      },
    );

    wrapper.find('#confirmpassword').simulate(
      'change', {
        target: {
          id: 'confirmpassword', value: 'sameWhiteLady',
        },
      },
    );

    wrapper.find('#gender').simulate(
      'change', {
        target: {
          id: 'gender', value: 'Male',
        },
      },
    );
    const file = new Blob([{
      lastModified: 1551810827095,
      name: 'dddd.png',
      size: 3516,
      type: 'image/png',
      webkitRelativePath: '',
    }]);
    wrapper.find('#profilepics').simulate(
      'change', {
        target: {
          id: 'profilepics',
          files: [file],
        },
      },
    );
    axios.put.mockImplementation(() => Promise.resolve({
      data: { Success: true, Message: 'profile updated successfully' },
    }));
    wrapper.find('#updateProfileButton').simulate(
      'click',
      {
        preventDefault: () => {},
      },
    );

    expect(dispatchMock).toHaveBeenCalled();
  });


  it('should throw error there is an empty input field', () => {
    wrapper.find('#phoneno').simulate(
      'change', {
        target: {
          id: 'phoneno', value: '07087653446',
        },
      },
    );

    wrapper.find('#password').simulate(
      'change', {
        target: {
          id: 'password', value: 'sameWhiteLady',
        },
      },
    );

    wrapper.find('#confirmpassword').simulate(
      'change', {
        target: {
          id: 'confirmpassword', value: 'sameWhiteLady',
        },
      },
    );

    wrapper.find('#gender').simulate(
      'change', {
        target: {
          id: 'gender', value: '',
        },
      },
    );
    const file = new Blob([{
      lastModified: 1551810827095,
      name: 'dddd.png',
      size: 3516,
      type: 'image/png',
      webkitRelativePath: '',
    }]);
    wrapper.find('#profilepics').simulate(
      'change', {
        target: {
          id: 'profilepics',
          files: [file],
        },
      },
    );
    const error = {
      response: {
        data: {
          Message: 'No empty field is allowed',
        },
      },
    };
    axios.put.mockImplementation(() => Promise.reject(error));
    wrapper.find('#updateProfileButton').simulate(
      'click',
      {
        preventDefault: () => {},
      },
    );

    expect(dispatchMock).toHaveBeenCalled();
  });

  it('should setup default state values', () => {
    const state = updateProfileReducer(undefined, {
      type: '@@INIT',
    });
    expect(state).toEqual({
      pending: false,
      error: '',
      success: false,
      message: '',
    });
  });

  it('should update state when action type is  UPDATE_IN_PROGRESS', () => {
    const state = updateProfileReducer(undefined, {
      type: 'UPDATE_IN_PROGRESS',
    });
    expect(state).toEqual({
      ...state,
      pending: true,
    });
  });

  it('should update state when action type is  UPDATE_SUCCESS', () => {
    const state = updateProfileReducer(undefined, {
      type: 'UPDATE_SUCCESS',
      payload: {
        success: true,
      },
    });
    expect(state).toEqual({
      ...state,
      success: true,
    });
  });

  it('should update state when action type is  UPDATE_ERROR', () => {
    const state = updateProfileReducer(undefined, {
      type: 'UPDATE_ERROR',
      error: '',
    });
    expect(state).toEqual({
      ...state,
      error: '',
      pending: false,
    });
  });
});
