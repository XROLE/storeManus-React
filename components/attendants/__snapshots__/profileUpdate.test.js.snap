// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component Should render just the Attendants component 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
